src/metabase/upload.clj

```diff
@@ -603,100 +600,132 @@
-                                        (= relaxed-types new-column-types))
-                               (let [fields (map normed-name->field normed-header)]
-                                 (->> (changed-field->new-type fields old-column-types relaxed-types)
-                                      (alter-columns! driver database table))))
-          ;; this will fail if any of our required relaxations were rejected.
-          parsed-rows        (parse-rows settings new-column-types rows)
-          row-count          (count parsed-rows)]
+  (try
+    (with-open [reader (bom/bom-reader file)]
+      (let [timer              (start-timer)
+            [header & rows]    (without-auto-pk-columns (csv/read-csv reader))
+            driver             (driver.u/database->driver database)
+            normed-name->field (m/index-by (comp normalize-column-name :name)
+                                           (t2/select :model/Field :table_id (:id table) :active true))
+            normed-header      (map normalize-column-name header)
+            create-auto-pk?    (and
+                                (driver/create-auto-pk-with-append-csv? driver)
+                                (not (contains? normed-name->field auto-pk-column-name)))
+            _                  (check-schema (dissoc normed-name->field auto-pk-column-name) header)
+            settings           (upload-parsing/get-settings)
+            old-column-types   (map (comp base-type->upload-type :base_type normed-name->field) normed-header)
+            ;; in the happy, and most common, case all the values will match the existing types
+            ;; for now we just plan for the worst and perform a fairly expensive operation to detect any type changes
+            ;; we can come back and optimize this to an optimistic-with-fallback approach later.
+            relaxed-types      (relax-types settings old-column-types rows)
+            new-column-types   (map #(if (matching-or-upgradable? %1 %2) %2 %1) old-column-types relaxed-types)
+            _                  (when (and (not= old-column-types new-column-types)
+                                          ;; if we cannot coerce all the columns, don't bother coercing any of them
+                                          ;; we will instead throw an error when we try to parse as the old type
+                                          (= relaxed-types new-column-types))
+                                 (let [fields (map normed-name->field normed-header)]
+                                   (->> (changed-field->new-type fields old-column-types relaxed-types)
+                                        (alter-columns! driver database table))
+                                   (->> (added-field->new-type normed-header old-column-types relaxed-types)
+                                        (add-columns! driver database table))))
```

Comment:
I read this a few times and figured that `changed-field->new-type` and `added-field->new-type` are mostly doing the same thing, are both complicated by the fact that they need to handle both new fields and existing fields. Splitting them out like the following should remove the need for them.

(I haven't tested this at all, and it most definitely doesn't work. Treat it like pseudocode.)

```clojure
(let [{new-fields      false
       existing-fields true}
      (group-by (comp some? :old-type)
                (map (fn [x y z] {:name x :old-type y, :new-type z})
                     normed-header
                     old-column-types
                     relaxed-types))]
  (alter-columns! driver database (into {}
                                        (for [{:keys [name old-type new-type]} existing-fields
                                              :when (not= old-type new-type)]
                                          [name new-type])))
  (add-columns! driver database (into {}
                                      (for [{:keys [name new-type]} new-fields]
                                        [name new-type]))))
```

Reply:
Perhaps this whole section could be extracted into a helper function, since this function is getting quite large.

Reply:
Yeah, calculating both change sets together makes a lot of sense, it avoids the "update" logic needing to explicitly check that something isn't new, and lets us deal with the awkward triple-iteration once.

I've gone with a slightly different implementation in the next revision, but taken inspiration.

Reply:
Resolving this orphaned thread, but feel free to pick it up on the new code.